### PR TITLE
Slasher span cache

### PR DESCRIPTION
**Pull request by shayzluf**
[Part of] #3133

---

# Description
detecting Slashing events is a time sensitive process . If we fail to do it fast enough we are going to start getting attestations backlogs that will add up quickly.
Slasher should be able to detect at least 20K attestations (in an aggregated form) in order to start being used in runtime.
This PR is aiming to do the following:
- [x] FRU cache for span maps
- [x] on FRU eviction save to db
- [x] on load re-detect all attestations in db to update span-maps
- [x] lazy load all db to cache
- [x] save all cached items to db on close

**Write why you are making the changes in this pull request**
While developing a tool to detect all previous attestations in testnet db detection performance wasn't good enough. 50ms much of that time was spent on saving and retrieving span maps

### CODE REVIEW

File: slasher/db/db.go [hunk 1]
```diff
@@ -35,8 +44,10 @@ func (db *Store) view(fn func(*bolt.Tx) error) error {
 }
 
 // NewDB initializes a new DB.
-func NewDB(dirPath string) (*Store, error) {
-	return NewKVStore(dirPath)
+func NewDB(dirPath string, ctx *cli.Context) (*Store, error) {
```
[the remark this answers is not in the document]
Comment (shayzluf) [reply slasher/db/db.go]: for doing this `if db.ctx.GlobalBool(flags.UseSpanCacheFlag.Name) {` later

File: slasher/db/min_max_span.go [hunk 2]
```diff
@@ -1,13 +1,38 @@
...
 	"github.com/pkg/errors"
 	slashpb "github.com/prysmaticlabs/prysm/proto/slashing"
 	"github.com/prysmaticlabs/prysm/shared/bytesutil"
+	"github.com/prysmaticlabs/prysm/slasher/flags"
 )
 
+var highestValidatorIdx uint64
+
+func saveToDB(validatorIdx uint64, _ uint64, value interface{}, cost int64) {
+	log.Infof("evicting span map fro validator id: %d", validatorIdx)
```
[the remark this answers is not in the document]
Comment (shayzluf) [reply slasher/db/min_max_span.go]: changed to trace. thanks

[the remark this answers is not in the document]
Comment (shayzluf) [reply slasher/db/min_max_span.go]: fixed. thanks

File: slasher/db/db.go [hunk 3]
```diff
@@ -5,18 +5,27 @@ import (
...
+
 	"github.com/boltdb/bolt"
 	"github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
 )
 
 var log = logrus.WithField("prefix", "slasherDB")
+var cacheItems = int64(20000)
+var maxCacheSize = int64(2 << 30) //(2GB)
+var spanCache *ristretto.Cache
```
[the remark this answers is not in the document]
Comment (shayzluf) [reply slasher/db/db.go]: moved to store

File: WORKSPACE [hunk 4]
```diff
@@ -1473,3 +1473,9 @@ go_repository(
     commit = "d7df74196a9e781ede915320c11c378c1b2f3a1f",
     importpath = "github.com/cespare/xxhash",
 )
+
+go_repository(
```
[the remark this answers is not in the document]
Comment (shayzluf) [reply WORKSPACE]: removed

File: slasher/db/db.go [hunk 5]
```diff
@@ -5,18 +5,27 @@ import (
 	"path"
 	"time"
 
+	"github.com/dgraph-io/ristretto"
+
```
Comment (nisdas) [inline slasher/db/db.go]:
```suggestion
```

File: slasher/db/db.go [hunk 6]
```diff
@@ -5,18 +5,27 @@ import (
...
+	"github.com/urfave/cli"
+
```
Comment (nisdas) [inline slasher/db/db.go]:
```suggestion
```

File: slasher/db/db.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (nisdas) [reply slasher/db/db.go]: the correct way is to use a config, if you want to inject any booleans from flags. We shouldnt be passing in the cli context

File: slasher/db/db.go [hunk 7]
```diff
@@ -76,8 +87,23 @@ func NewKVStore(dirPath string) (*Store, error) {
...
+	}
+	spanCache, err := ristretto.NewCache(&ristretto.Config{
+		NumCounters: cacheItems,   // number of keys to track frequency of (10M).
+		MaxCost:     maxCacheSize, // maximum cost of cache.
+		BufferItems: 64,           // number of keys per Get buffer.
+		OnEvict:     saveToDB,
+	})
+	if err != nil {
+		errors.Wrap(err, "failed to start span cache")
+		panic(err)
```
Comment (terencechain) [inline slasher/db/db.go]: Not sure about this panic, if you look at how we initialize beacon node db. Panic is not used there

File: slasher/db/db.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (shayzluf) [reply slasher/db/db.go]: thanks. fixed

File: slasher/db/db.go (hunk 6 above)
Comment (shayzluf) [reply slasher/db/db.go]: fixed

File: slasher/db/db.go (hunk 7 above)
Comment (shayzluf) [reply slasher/db/db.go]: changed to return error

Comment (shayzluf) [reply slasher/db/db.go]: thanks

File: slasher/service/service.go [hunk 8]
```diff
@@ -176,6 +179,25 @@ func (s *Service) startSlasher() {
 	}()
 }
 
+func (s *Service) loadSpanMaps(err error, slasherServer rpc.Server) {
+	lt, err := slasherServer.SlasherDB.LatestIndexedAttestationsTargetEpoch()
+	if err != nil {
+		log.Errorf("Could not extract latest target epoch from indexed attestations store:%v", err)
```
Comment (terencechain) [inline slasher/service/service.go]:
```suggestion
		log.Errorf("Could not extract latest target epoch from indexed attestations store: %v", err)
```

File: slasher/service/service.go [hunk 9]
```diff
@@ -176,6 +179,25 @@ func (s *Service) startSlasher() {
...
+	}
+	for i := uint64(0); i < lt; i++ {
+		ias, err := slasherServer.SlasherDB.IndexedAttestations(i)
+		if err != nil {
+			log.Errorf("Got error while trying to retrieve indexed attestations from db:%v", err)
```
Comment (terencechain) [inline slasher/service/service.go]:
```suggestion
			log.Errorf("Got error while trying to retrieve indexed attestations from db: %v", err)
```

File: slasher/service/service.go [hunk 10]
```diff
@@ -176,6 +179,25 @@ func (s *Service) startSlasher() {
...
+		}
+		for _, ia := range ias {
+			slasherServer.UpdateSpanMaps(s.context, ia)
+		}
+		if i%5 == 0 {
```
Comment (terencechain) [inline slasher/service/service.go]: why 5? can you add a comment?

File: slasher/service/service.go [hunk 11]
```diff
@@ -176,6 +179,25 @@ func (s *Service) startSlasher() {
...
+			log.Infof("update span maps for epoch: %d", i)
```
Comment (terencechain) [inline slasher/service/service.go]:
```suggestion
			log.Infof("Update span maps for epoch: %d", i)
```

If this happens a lot, feel free to switch to debug..

File: slasher/service/service.go [hunk 12]
```diff
@@ -233,6 +255,9 @@ func (s *Service) Close() {
 	if err != nil {
 		log.Panicf("Could not stop the slasher service: %v", err)
 	}
+	if err := s.slasherDb.SaveCachedSpansMaps(); err != nil {
+		log.Fatal("didnt save span map cache to db. if span cache is enabled please restart with --%s", flags.RebuildSpanMapsFlag.Name)
```
Comment (terencechain) [inline slasher/service/service.go]:
```suggestion
		log.Fatal("Didn't save span map cache to db. if span cache is enabled please restart with --%s", flags.RebuildSpanMapsFlag.Name)
```

File: slasher/service/service.go (hunk 10 above)
Comment (shayzluf) [reply slasher/service/service.go]: removed. did it to reduce clutter in the terminal

**Pull request merged**